README.MD
# SERVEUR MÉDIA SÉCURISÉ POUR COLLABORATEURS DU DOMUS

## Apperçu

- Permet le streaming de médias compatibles avec les formats mime html5 [définis dans cette section](#mime_types).
- La configuration se fait à partir du fichier config.json, à la racine du repo.
- Permet de "matcher" des fichiers à partir d'expressions régulières et d'associer [les routes média correspondantes](#medias).
- Léger et souple. Implémenté en javascript/coffeescript (nodejs)
- Sécurisé, avec [identification par tokens](#auth)
- Gestion des droits d'accès par
    - [application cliente](#applications)
    - [projet](#projects)
- Interfaçage avec n'importe quel mécanisme d'authentification, il suffit [d'écrire un petit module](#module)

## Conventions
 `$uneVariable` fait référence, suivant le contexte
- à un champs (json) du fichier de configuration, ex : `$uneVariable` fait référence à `"uneVariable":"un_string"`
- à une variable de route HTTP, ex : pour la route `/a/$uneVariable/`, et le chemin   `/a/uneValeurDeVariable/` `$uneVariable` fait référence à `"uneValeurDeVariable"`


## Intégration, configuration

<a name="applications"></a>
### Configuration d'une application ( $applications )
Pour clairement définir les accès à différents médias, le serveur permet la déclaration d'applications.
Les clients de ces applications auront des droits spécifiques, et des mécanismes d'authentification propres.
En voici un exemple :

```json
    "AppName":{
      "auth_handler":{
        "name": "some_auth_handler",
        "uri" :" some_data_store://some_address:some_port",
        "options":{},
        "reconnectAfter_s":6000
      },
      "projects":["ProjectName1","ProjectName2"],
      "enableMeta":true
    }
  }
```
avec :
- `$AppName`      : le nom de l'application, utilisée par la suite dans toutes les requêtes
- `$auth_handler` : le gestionnaire d'authentification, cf [la section dédiée](#module)
  - `$name`       : le nom du gestionnaire d'authentification utilisé, parmis ceux déclarés dans $auth_handlers
  - `$uri`        : le universal resource identifier qui permettra à au gestionnaire de se connecter à la bonne adresse
  - `$options`     : des options spécifiques au gestionnaire
  - `$reconnectAfter_s` : durée en secondes après laquelle media-node s'autorise à relancer le gestionnaire s'il est défaillant (entrepot injoignable par exemple).
- `$projects`     : les noms des projets auxquels auront accès les clients de l'application
- `$enableMeta`    : autorise la diffusion [de meta-informations](#meta-infos) lors des [place requests](#request_place) si et seulement si le projet a l'option `$supportsMeta` activée.


<a name="projects"></a>
### Configuration d'un projet ( $projects )
Un "projet" du point de vue du serveur est un simple répertoire racine.
Il peut être configuré ainsi , dans l'entrée  `projects` du fichier de configuration :

```json
    "ProjectName":{
      "rootDir":"my_project_dir",
      "expRegex":"*",
      "mediaDir":"videos/"
      "mediaRegex":"*.mp4",
      "supportsMeta":true
    }
```
avec :

- `$ProjectName` : le nom du projet, utilisé dans les requêtes  
- `$rootDir` : le répertoire racine, relatif à `$baseDir` du fichier de configuration. **NE DOIT NI COMMENCER, NI SE TERMINER PAR /**
- `$expRegex` : l'epression régulière pour matcher les répertoires propres à une expérimentation à partir de `$rootDir` (descendants directs).
- `$mediaDir` : le chemin permettant d'atteindre les médias depuis `$expdir`, peut être vide `""`,null ou non renseigné. **NE DOIT NI COMMENCER, NI SE TERMINER PAR /**
- `$mediaRegex`: l'expression régulière pour matcher les médias à partir de `$expDir` et `placeDir`
- `$placeDir` : n'est pas configurable. Passé en paramètre d'une requête. Chacun de ces dossiers permets de distinguer différents médias suivant leur positionnement. (par exemple `cuisine`, `salle-manger` ...)
- `$supportsMeta` : signale que [les meta-infos par fichier YAML](#meta-infos) sont activées si sa valeur est `true`

**REMARQUE IMPORTANTE** : la syntaxe des regexp est spécifique à une librairie utilisée pour le projet, [**veuillez vous renseigner ici.**](https://github.com/Carrooi/Node-FsFinder#path-mask)

- une regexp est contenur entre crochets `<[0-9]+>`
- la wildcard `*` équivaut à `<[0-9a-zA-Z/.-_ ]+>`

### Configuration du serveur ( $serv )
 Voici un exemple :

 ```json

 "serv":{
  "port":5000,
  "baseDir":"/home/dummy/medias/CoLab/",
  "disableAuth":true,
  "debug":true,
  "logLevel":"warn",
  "delTokensOnShutdown":false,
  "streamer":{
     "maxAge":86400,
     "customMimes":{
       ".web-mp4":"video/mp4"
     }
   }
}
 ```  
 avec :

 - `$port` : le port d'écoute du serveur
 - `$baseDir` : le répertoire racine à partir duquel le serveur à accès. **Remarque** : seuls les répertoires appartenant au groupe `media-node`
 - `$disableAuth` : désactiver [le module d'authentification](#module). La route d'authentification ne sera pas disponible et les accès aux médias ne requèreront pas de token. **IMPORTANT** : n'est activable que lorsque `$debug`=true.
 - `$debug` : lancer le serveur en mode débug. Donne droit à l'activation de `$disableAuth`, et permet de rediriger les erreurs dans le corps de la réponse HTTP. Très pratique pour débugguer !
 - `$logLevel` : une valeur parmi `["error", "warn", "info", "debug", "trace"]`. Défaut : `"info"`
 - `$delTokensOnShutdown` : invalide les tokens lorsque le serveur s'arrête.
 - `$streamer.$customMimes` : Permet de mapper des associations de fichier avec [un mime-type](#mime_types) (défaut `{}`) .
 - `$streamer.$maxAge` : Indique au navigateur combien de secondes devrait-il conserver en câche le média téléchargé (défaut `3600`).

### Configuration des gestionnaires d'authentification ( $auth_handlers )

Confère [la section dédiée](#modules)

<a name="redis"></a>
### Configuration de Redis ( $redis )
**Redis doit être installé pour que le serveur fonctionne**
Exemple de configuration (config.json):
```json
"redis":{
  "socket":"/tmp/redis-media-node.sock",
  "expire_min":1440
}
```
Avec `$socket` le socket unix utilisé pour communiquer avec le video-node (ne devrait pas être modifié) et `$expire_min` la durée de vie d'un token en minutes.

## Dépendances

- Redis, v2.8 ou supérieur (minor releases)
- Node v0.10 et supérieurs
- npm v1.3 et supérieurs
- coffeescript v1.10 (**via npm** : `sudo npm install -g coffee-script@1.10` )

Exemple pour **debian, ubuntu...**  
> sudo apt-get install redis-server nodejs npm

**Problème connu avec nodejs**  
Pour certaines verison de node, il y a un conflit avec un packet node qui n'a rien à voir avec nodejs.  
Faites `whereis node`.
**Si aucun répertoire n'apparait**, il faut faire un symlink de nodejs vers node :  
> sudo ln -s /usr/bin/nodejs /usr/sbin/node
> sudo ln -s /usr/bin/nodejs /usr/bin/node


<a name="API"></a>
## API

<a name="overview"></a>
**APPERÇU DE L'IMPLÉMENTATION d'UN ACCES A UN MDEDIA** : Lorsqu'une requête d'accès à un média est envoyée au serveur ([cf section API](#API)), la démarche pour diffuser le bon fichier va être la suivante.
On suppose que cette requête contient : `$ProjectName`, `$ExpName`, et `$placeDir`

1. Les paramètres `a` (application) et `t` existent-t-ils? Si oui, étape 2, sinon erreur http `422` [avec identifiant d'erreur "SESSION.BAD_QUERY"](#error_ids)
2. L'application est-elle enregistrée? Vérification avec le paramètre `a`. Si oui, étape 3, sinon erreur http `403` [avec identifiant d'erreur "INTERNAL.APPLICATION_NOT_REGISTERED"](#error_ids)
3. Le client est-il autorisé? Vérification avec le paramètre `t` (token). Si oui, étape 6, sinon, erreur http `403` [avec identifiant d'erreur "SESSION.FORBIDDEN"](#error_ids)
3. Le projet est-t-il configuré? Si oui, étape 4, si non, erreur http `500` [avec identifiant d'erreur "INTERNAL.PROJECT_NOT_CONFIGURED"](#error_ids)
4. La racine du projet est-elle configurée? Si oui, étape 4, si non, erreur http `500` [avec identifiant d'erreur "INTERNAL.PROJECT_ROOT_MISSING"](#error_ids)
5. L'application a-t-elle le droit d'accéder au projet `$ProjectName` ? Si oui, étape 5, sinon, erreur http `403` [avec identifiant d'erreur "SESSION.PROJECT_ACCESS_FORBIDDEN"](#error_ids)
6. Le répertoire associé au projet est-il accessible? Si oui, étape 7, si non, erreur http `404` [avec identifiant d'erreur "SESSION.ROOT_DIR_NOT_FOUND"](#error_ids)
7. Le dossier associé à expérimentation `$ExpName`, matché avec `$expRegex$exp` existe-t-elle? Si oui, étape 8, si non, erreur http `404` [avec identifiant d'erreur "SESSION.EXP_DIR_NOT_FOUND"](#error_ids)
8. Le dossier associé à l'endroit de captation `$placeDir` existe-t-il? Si oui, étape 9, si non erreur http `404` [avec identifiant d'erreur "SESSION.PLACE_DIR_NOT_FOUND"](#error_ids)
9. Une $média matchée par `$mediaRegex` est-elle matchée à l'intérieur de `$placeDir`? Si oui, étape 10, si non erreur http `404` [avec identifiant d'erreur "SESSION.MEDIA_NOT_FOUND"](#error_ids)
10. Renvoi d'une réponse http avec status `200`, `Content-Type : $type` ou `$type` est un [parmi la liste suivante](#mime_types) et évidemment le contenu du média dans le body.

<a name="error_ids"></a>
### Codes d'erreurs
Ces codes d'erreurs sont distincts d'une erreur http.
Le serveur va renvoyer une réponse de format mime `application/json` contenant
un string identifiant l'erreur rencontrée. Ce string peut être récupéré via JSON.parse(response) en javascript par exemple.  
**Un code d'erreur peut donc très bien s'ajouter à une erreur http!**

| Code                                  |  Signification |
|:--------------------------------------|:---------------|
| "INTERNAL.APPLICATION_NOT_REGISTERED" | L'application n'est pas enregistrée |
| "INTERNAL.REDIS_INTERNAL_ERROR"       | Redis (db de cache) a rencontré un problème |
| "INTERNAL.PROJECT_NOT_CONFIGURED"     | Le projet n'est pas configuré sur le serveur (aucune entrée dans `projects` correspondante) |
| "INTERNAL.PROJECT_ROOT_MISSING"       | Le projet est configuré, mais la propriété "rootDir" est manquante |
| "AUTH_FAILURE.HANDLER_INTERNAL_ERROR" | Le gestionnaire d'authentification a rencontré une erreur, probablement en contactant l'entrepot distant |
| "AUTH_FAILURE.BAD_CREDENTIALS"        | Le gestionnaire d'authentification a indiqué que les 'credentials' n'avaient pas le format escomptée |
| "AUTH_FAILURE.FORBIDDEN"              | Le gestionnaire d'authentification a indiqué que les 'credentials' étaient invalides |
| "SESSION.FORBIDDEN"                   | L'identification par token a échoué (token expiré, token invalide)  |
| "SESSION.APP_MISSING_PREROGATIVES"    | Le projet n'est pas référencé dans la configuration de l'application |
| "SESSION.BAD_QUERY"                   | La arguments 'query' sont manquants ou mal formés  |
| "SESSION.ROOT_DIR_NOT_FOUND"          | Le répertoire racine du projet n'a pas été trouvé |
| "SESSION.EXP_DIR_NOT_FOUND"           | Le répertoire d'expérimentation n'a pas été trouvé |
| "SESSION.MEDIA_DIR_NOT_FOUND"         | Le répertoire contenants les médias (`$mediaDir` dans la configuration d'un projet) n'a pas été trouvé |
| "SESSION.PLACE_DIR_NOT_FOUND"         | Le répertoire associé à l'endroit de captation (ex `cuisine`) n'a pas été trouvé |
| "SESSION.MEDIA_NOT_FOUND"             | Aucun média n'a été trouvée dans le répertoire associé à l'endroit de captation. |

### Requêtes
Les requêtes sont évaluées par routes ([path](http://tools.ietf.org/html/rfc3986#section-3.3)).
**Si l'authentification est activée, elles doivent contenir des paramètres d'identification, cf [cette section](#identification)**

<a name="request_place"></a>
#### [REQUEST_PLACES] Récupération de la liste des endroits de captation (places)
Cette requête renvoie :

**Si le client a le droit, et le projet correctement configuré, une réponse avec status `200` sous format application/json au format utf-8**

- Un objet dont l'attribut `places` est une liste sous forme de tableau des endroits de captation trouvé pour une expérimentation donnée si le dossier d'expérimentation existe. **ATTENTION** Le serveur ne vérifie pas si ces dossiers contiennent un média. Chaque endroit de captation est un objet contenant les champs suivants:
  - `place` : le lieu de captation, c-à-d le nom du dossier
  - `fileFound` : `true` si un fichier a été trouvé, `false` sinon
  - `meta` : **optionnel** Si l'option `$supportsMeta` est activée pour le projet **et** l'option `$enableMeta` est activée pour l'application :
    - un objet contenant les champs [du fichier YAML](#meta-infos) si un tel fichier existe
    - `{}` sinon
- Un [code d'erreur](#error_ids) si un problème est rencontré

Exemple :
```json
{
  "places": [
    {
      "place": "chambre",
      "fileFound": true,
      "meta": {
        "warning": "Ne fonctionne plus a partir de la minute 32",
        "comments": "Nous avertir s'il manque une partie de video que vous trouvez importante.",
        "soundQ": 2
      }
    },
    {
      "place": "cuisine-plafond",
      "fileFound": true,
      "meta": {}
    }
  ]
}
```

**Sinon**

- Une erreur http `500` avec [code d'erreur  "INTERNAL.PROJECT_NOT_CONFIGURED"](#error_ids)
- ou Une erreur http `403` avec [code d'errreur "SESSION.FORBIDDEN"](#error_ids)


Elle a la forme suivante : `/i/$ProjectName/$ExpName` ou  :

- `$ProjectName` correspond au `$ProjectName` du fichier de configuration
- `$ExpName` est le nom de l'expérimentation. ça n'est pas le nom complêt puisque le dossier va être matché avec "$expRegex$ExpName"

<a name="medias"></a>
#### [REQUEST_MEDIA] Récupération d'un media

**REMARQUE IMPORTANTE** : il est très recommandé de faire une [REQUEST_PLACES](#request_place) d'abord, pour s'assurer des  endroits de captation disponibles.

**Si le client a le droit, et le projet correctement configuré**

- Si le media est trouvé, une réponse avec status `200` et `Content-Type : $type` ou `$type` est un [parmi la liste suivante](#mime_types)
- Sinon, une réponse avec status `200` et `Content-Type : application/json` contenant un [code d'erreur](#error_ids)

**Sinon**

- Une erreur http `500` avec [code d'erreur  "INTERNAL.PROJECT_NOT_CONFIGURED"](#error_ids)
- ou Une erreur http `500` avec [code d'erreur  "INTERNAL.PROJECT_NOT_CONFIGURED"](#error_ids)
- ou Une erreur http `403` avec [code d'errreur "SESSION.FORBIDDEN"](#error_ids)

Elle a la forme suivante : `/v/$ProjectName/$ExpName/$placeDir` ou  :

- `$ProjectName` correspond au `$ProjectName` du fichier de configuration
- `$ExpName` est le nom de l'expérimentation. ça n'est pas le nom complêt puisque le dossier va être matché avec "$expRegex$ExpName"
- `$placeDir` est le nom de l'endroit de captation (`cuisiniere`,`salle-de-bain` ...)

[Confère cette description pour connaitre le détail de l'implémentation.](#overview)

## Sécurité
<a name="auth"></a>
### Authentification par 'credentials'

Si l'authentification est activée ( `$disableAuth = false` dans le fichier de configuration ), le client va devoir s'authentifier auprès du serveur vidéo.
**Il est important de noter que le mécanisme d'authentification est à la charge du développeur souhaitant utiliser le serveur vidéo**.
Il devra écrire un petit module en javascript de quelques lignes permettant de valider les 'credentials' du client  
( typiquement en contactant la base de donnée utilisée par l'application hôte du développeur qui souhaite intégrer les vidéos ), voire [la section dédiée pour l'implémentation](#module).

<a name="auth_req"></a>
#### Requête (POST HTTP/1.1)

Il faut envoyer une requête :

- `POST`
- sur le chemin `/auth/c/$applicationName`
- de `Content-Type` mime `application/x-www-form-urlencoded`
- dont le corps (`body`) contient les 'credentials' [url-encodés](https://en.wikipedia.org/wiki/Percent-encoding#The_application.2Fx-www-form-urlencoded_type) qui seront transmis au [gestionnaire d'authentification](#module)

avec  :

- `$applicationName` le nom de l'application déclarée de le fichier de configuration JSON,
- 'credentials' les paramètres d'authentification qui seront passés au gestionnaire d'authentification de l'application (`$auth_handler` dans le fichier de configuration), une fois désérialisés.

**Remarque** : la raison du mime type `application/x-www-form-urlencoded` est que
les navigateurs implémentent aujourd'hui des politiques de sécurité contraignantes, les CORS, qui interdisent des requêtes `POST` de mime type `application/json` sur des domaines croisés.
Cf [cet article du MDN pour référence](https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS)

#### Réponse HTTP/1.1
La réponse de de type mime `application/json`

**Si l'authentification a réussie :** Renvoie un objet json contenant deux `string`s :

- l'un nommé **token**`, à conserver pour l'identification des requêtes  (cf section suivante)
- l'autre nommé **epoch_s**, contenant la date d'expiration du token en [temps epoch (secondes)](https://en.wikipedia.org/wiki/Unix_time)  

**Si l'authentification a échouée :** Renvoie un code d'erreur au format json dont le préfixe est `AUTH_FAILURE`, [confère la section dédiée pour en connaître la signification](#error_ids).

<a name="auth_token"></a>

### Authentification par token
Si un token est encore valide alors que le client relance l'application web utilisatrice du serveur vidéo, il n'est pas nécessaire de se
réauthentifier par 'credentials'. Pour s'assurer que le token est toujours valide côté serveur (qu'il n'y a pas eu de redémarrage brutal ou un problème rencontré par redis par exemple), une requête peut être envoyée.

#### Requête (POST HTTP/1.1)
La requête doit être :
- `POST`
- sur le chemin `/auth/t/$applicationName`
- de `Content-Type` mime `application/x-www-form-urlencoded`
- dont le corps (`body`) contient un paramêtre `token=mytoken` [url-encodés](https://en.wikipedia.org/wiki/Percent-encoding#The_application.2Fx-www-form-urlencoded_type) dont le serveur va s'assurer de la validité en contactant l'entrepot redis.

avec
- `$applicationName` le nom de l'application cliente.
- `my_token` le token renvoyé précédement par une authentification par 'credentials'.

#### Réponse HTTTP/1.1

Si le token est valide :
- une réponse vide avec status=200 (OK)
Sinon
- une réponse avec [code d'erreur "SESSION.FORBIDDEN"](#error_ids) et status=403

<a name="identification"></a>
### Identification

Une requête sur un chemin media `/m/*` ou info `/i/*` doit nécessairement contenir deux paramètres d'identification, **a** (application) et **t** (token)
Exemple d'url complète : `http://serveur-media/m/MY_PROJECT/MY_EXP/MY_PLACE?a=my_app&t=my_token`
avec `my_app` le nom de l'application enregistrée dans le fichier de configuration et `my_token` le token obtenu [à l'authentification](#auth).

<a name="module"></a>
## Modules / Gestionnaires d'authentification
Idéalement, les modules devraient être chargés avec npm. Mais nous n'avons pas de dépot npm propre au laboratoire...

Un module est un simple fichier javascript **ou** coffeescript qui a pour role d'authentifier un client.



### Prérequis

- Le module doit être définit dans le dossier `auth_handler`
- Dans un fichier `mon_module.js` ou `mon_module.coffee`
- Le nom du fichier doit être déclaré dans le fichier de configuration, tableau `$auth_handlers`, sans extension.
- **ses dépendances** doivent être déclarées dans le fichier de configuration de npm, `package.conf`, paramètre `$dependencies`. [Plus d'infos sur semver (la syntaxe pour déclarer les versions candidates) ici](http://semver.npmjs.com/).

### Conseils
- Les dépendances typiques seront des drivers pour l'entrepot de données utilisé. Faites une recherche "npm postgresql" dans google par exemple pour trouver un driver postegresql.
- Utilisez [lodash](https://lodash.com/), qui est déjà incluse dans le projet, une excellente lib fonctionnelle.


<a name="handlers">
### Spécifications

Rien de mieux qu'un exemple javascript :
**REMARQUES IMPORTANTES**

- aucune gestion d'erreur à implémenter, elles sont faites en amont
- l'utilisation de `this.logger` est fortement encouragée pour journaliser des informations et/ou debugger. [Informations sur la librairie de logging utilisée ici](https://www.npmjs.com/package/pince)

```javascript
var postgres=require("node-postgres"); // juste pour l'exemple, mais ça peut être n'importe quelle package npm
// Du code pour déclarer des fonctions privées etc...

// exemple d'uri : "postgres://username:password@localhost/database"
// options est un objet javascript
// Promise, le package promise, cf https://www.npmjs.com/package/promise
module.exports = function(uri,options,Promise){
    // DOIT renvoyer un objet ayant EXACTEMENT (ni plus ni moins) les méthodes suivantes
    return {
        assertCredentials:function(credentials){

HISTORY.MD
v0.3.2

- fix bug majeur lorsqu'une erreur est rencontrée par le serveur (mauvais code de status http
- fix bug majeur lorsqu'un fichier meta n'est pas disponible à la racine du projet²

v0.3.1

- fix dans le script d'installation : vérification de l'existance de la commande redis-server
Z
v0.3.0

- gestion du fichier de meta par défaut à la racine d'un projet


v0.2.2 

- passage du package *promise* en paramètre d'un gestionaire d'authentification

v0.2.1 
 
- amélioration de la gestion d'erreurs  du gestionnaire mongo

v0.2.0 (First release)

- Compatible avec **CoLab** 0.2
- Implémentation du auth_handler mongo pour CoLab

v0.1.0 (draft)

- définition des routes serveur via express
- mécanisme d'authentification délégué (voir auth_handler dans la documentation)
- attribution de jetons (tokens) aux utilisateurs authentifiés donnant un accès aux médias
- routage intelligent avec déduction de routes par regex
- découverte de médias à partir d'une structure de dossier conventionnelle 
  - répertoire racine d'un projet
  - répertoire racine des expérimentations
  - répertoire racine d'une expérimentation
  - répertoire racine des médias propres à l'expérimentations
  - répertoire racine d'un point de captation
- serialisation des chemins dans une base redis pour optimiser le temps de requêtes médias multiples
  


lib/vidStreamer.js
/*!
* VidStreamer.js
*
* Copyright (c) 2012 Andrew Weeks http://meloncholy.com
* Licensed under the MIT licence. See http://meloncholy.com/licence
* Version 0.1.4
*/

"use strict";

var fs = require("fs");
var url = require("url");
var events = require("events");
var settings = require('./streamer-conf');
var Throttle = require('throttle');
var _        = require("lodash");

var handler = new events.EventEmitter();

// Stuff to serve. Don't add null or "null" to the list (".null" should be fine) as the regex extension check will fail and you'll have a big security hole. And obviously don't add .js, .php or anything else you don't want to serve as source either.
var mimeTypes = _.extend({
	".swf": "application/x-shockwave-flash",
	".flv": "video/x-flv",
	".f4v": "video/mp4",
	".f4p": "video/mp4",
	".mp4": "video/mp4",
	".asf": "video/x-ms-asf",
	".asr": "video/x-ms-asf",
	".asx": "video/x-ms-asf",
	".avi": "video/x-msvideo",
	".mpa": "video/mpeg",
	".mpe": "video/mpeg",
	".mpeg": "video/mpeg",
	".mpg": "video/mpeg",
	".mpv2": "video/mpeg",
	".mov": "video/quicktime",
	".movie": "video/x-sgi-movie",
	".mp2": "video/mpeg",
	".qt": "video/quicktime",
	".mp3": "audio/mpeg",
	".wav": "audio/x-wav",
	".aif": "audio/x-aiff",
	".aifc": "audio/x-aiff",
	".aiff": "audio/x-aiff",
	".jpe": "image/jpeg",
	".jpeg": "image/jpeg",
	".jpg": "image/jpeg",
	".png" : "image/png",
	".svg": "image/svg+xml",
	".tif": "image/tiff",
	".tiff": "image/tiff",
	".gif": "image/gif",
	".webm": "video/webm",
	".m3u8": "application/x-mpegurl",
	".ts": "video/mp2t",
	".ogg": "video/ogg"
},settings.customMimes||{});

var vidStreamer = function (req, res, path,fsErrorCbk) {
	var stream;
	var stat;
	var info = {};
	var ext;
	var range = typeof req.headers.range === "string" ? req.headers.range : undefined;
	var reqUrl = url.parse(req.url, true);

	info.path = path || (typeof reqUrl.pathname === "string" ? reqUrl.pathname.substring(1) : undefined);
	if (info.path) {
		try {
			info.path = decodeURIComponent(info.path);
		} catch (exception) {
			// Can throw URI malformed exception.
			handler.emit("badRequest", res);
			return false;
		}
	}

	// Security checks. Word.
	if (!info.path) {
		handler.emit("fileCheck", res, fsErrorCbk);
		return false;
	}
	else if (info.path.search(/^\.\.?|^\/|^\\/) !== -1) {
		handler.emit("security", res, { message: info.path });
		return false;
	}
	else if (info.path.substring(0, settings.rootPath.length) !== settings.rootPath) {
		// This will trigger if wrong slashes are used. Change?
		console.log(info.path+" not a substring of " + settings.rootPath)
		handler.emit("security", res, { message: info.path });
		return false;
	} else {
		ext = info.path.match(/.*(\..+?)$/);

		if (ext === null || ext.length !== 2 || (info.mime = mimeTypes[ext[1].toLowerCase()]) === undefined) {
			handler.emit("badMime", res, { message: info.path });
			return false;
		}
	}

	info.path = info.path.substring(settings.rootPath.length);
	info.file = info.path.match(/(.*[\/|\\])?(.+?)$/)[2];

	// Want that file or a consisten-but-random one instead (for demos)?
	if (settings.random) {
		if (!randomFile(info)) {
			handler.emit("noRandomFiles", res, { message: info.path });
			return false;
		}
	} else {
		info.path = settings.rootFolder + info.path;
	}

	try {
		stat = fs.statSync(info.path);

		if (!stat.isFile()) {
			handler.emit("fileCheck", res, fsErrorCbk);
			return false;
		}
	} catch (e) {
		handler.emit("fileCheck", res, fsErrorCbk, e);
		return false;
	}

	info.start = 0;
	info.end = stat.size - 1;
	info.size = stat.size;
	info.modified = stat.mtime;
	info.rangeRequest = false;

	if (range !== undefined && (range = range.match(/bytes=(.+)-(.+)?/)) !== null) {
		// Check range contains numbers and they fit in the file.
		// Make sure info.start & info.end are numbers (not strings) or stream.pipe errors out if start > 0.
		info.start = isNumber(range[1]) && range[1] >= 0 && range[1] < info.end ? range[1] - 0 : info.start;
		info.end = isNumber(range[2]) && range[2] > info.start && range[2] <= info.end ? range[2] - 0 : info.end;
		info.rangeRequest = true;
	} else if (reqUrl.query.start || reqUrl.query.end) {
		// This is a range request, but doesn't get range headers. So there.
		info.start = isNumber(reqUrl.query.start) && reqUrl.query.start >= 0 && reqUrl.query.start < info.end ? reqUrl.query.start - 0 : info.start;
		info.end = isNumber(reqUrl.query.end) && reqUrl.query.end > info.start && reqUrl.query.end <= info.end ? reqUrl.query.end - 0 : info.end;
	}

	info.length = info.end - info.start + 1;

	downloadHeader(res, info);

	// Flash vids seem to need this on the front, even if they start part way through. (JW Player does anyway.)
	if (info.start > 0 && info.mime === "video/x-flv") {
		res.write("FLV" + pack("CCNN", 1, 5, 9, 9));
	}
	stream = fs.createReadStream(info.path, { flags: "r", start: info.start, end: info.end });

	if (settings.throttle) {
		stream = stream.pipe(new Throttle(settings.throttle))
	}

	stream.pipe(res);
	return true;
};

vidStreamer.settings = function (s) {
	for (var prop in s) { settings[prop] = s[prop]; }
	return vidStreamer;
};

var randomFile = function (info) {
	var fileCode = 0;
	var fileParts;
	var randomSource;
	var randomPath;
	var randomFiles;
	var extn = info.path.match(/\..+?$/);

	// No extension given
	if (extn === null) return false;
	extn = extn[0];

	// See if there's a path
	fileParts = info.path.match(/(.*[\/|\\])?(.+?)$/);
	if (fileParts[1] === undefined) {
		randomPath = settings.rootFolder;
		randomSource = info.path;
	} else {
		randomPath = settings.rootFolder + fileParts[1];
		randomSource = fileParts[2];
	}

	if (randomPath.substr(-1) !== "/" && randomPath.substr(-1) !== "\\") randomPath += "/";

	for (var i = 0, len = randomSource.length; i < len; i++) {
		fileCode += randomSource.charCodeAt(i);
	}

	try {
		randomFiles = fs.readdirSync(randomPath);
	} catch (e) {
		return false;
	}
	randomFiles = randomFiles.filter(function (el) { return el.substr(-extn.length) === extn; });
	if (randomFiles.length === 0) return false;
	info.path = randomPath + randomFiles[fileCode % randomFiles.length];
	return true;
};

var downloadHeader = function (res, info) {
	var code = 200;
	var header;

	// 'Connection':'close',
	// 'Cache-Control':'private',
	// 'Transfer-Encoding':'chunked'

	if (settings.forceDownload) {
		header = {
			Expires: 0,
			"Cache-Control": "must-revalidate, post-check=0, pre-check=0",
			//"Cache-Control": "private",
			"Content-Type": info.mime,
			"Content-Disposition": "attachment; filename=" + info.file + ";"
		};
	} else {
		header = {
			"Cache-Control": "public; max-age=" + settings.maxAge,
			Connection: "keep-alive",
			"Content-Type": info.mime,
			"Content-Disposition": "inline; filename=" + info.file + ";",
			"Accept-Ranges": "bytes"
		};

		if (info.rangeRequest) {
			// Partial http response
			code = 206;
			header.Status = "206 Partial Content";
			header["Content-Range"] = "bytes " + info.start + "-" + info.end + "/" + info.size;
		}
	}

	header.Pragma = "public";
	header["Last-Modified"] = info.modified.toUTCString();
	header["Content-Transfer-Encoding"] = "binary";
	header["Content-Length"] = info.length;
    if(settings.cors){
        header["Access-Control-Allow-Origin"] = "*";
        header["Access-Control-Allow-Headers"] = "Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept";
    }
    header.Server = settings.server;
	header.Server = settings.server;
	
	res.writeHead(code, header);
};

var errorHeader = function (res, code) {
	var header = {
		"Content-Type": "text/html",
		Server: settings.server
	};

	res.writeHead(code, header);
};

// http://stackoverflow.com/a/1830844/648802
var isNumber = function (n) {
	return !isNaN(parseFloat(n)) && isFinite(n);
};

// A tiny subset of http://phpjs.org/functions/pack:880
var pack = function (format) {
	var result = "";

	for (var pos = 1, len = arguments.length; pos < len; pos++) {
		if (format[pos - 1] == "N") {
			result += String.fromCharCode(arguments[pos] >> 24 & 0xFF);
			result += String.fromCharCode(arguments[pos] >> 16 & 0xFF);
			result += String.fromCharCode(arguments[pos] >> 8 & 0xFF);
			result += String.fromCharCode(arguments[pos] & 0xFF);
		} else {
			result += String.fromCharCode(arguments[pos]);
		}
	}

	return result;
};

handler.on("fileCheck",function(res,fsErrorCbk,e){
	if(fsErrorCbk instanceof Function) fsErrorCbk();
	else handler.emit("badFile",res,fsErrorCbk||e);
});
handler.on("badFile", function (res, e) {
	errorHeader(res, 404);
	res.end("<!DOCTYPE html><html lang=\"en\">" +
		"<head><title>404 Not found</title></head>" +
		"<body>" +
		"<h1>Ooh dear</h1>" +
		"<p>Sorry, I can't find that file. Could you check again?</p>" +
		"</body></html>");
	console.error("404 Bad File - " + (e ? e.message : ""));
});

handler.on("badRange", function (res, e) {
	errorHeader(res, 416);
	res.end("<!DOCTYPE html><html lang=\"en\">" +
		"<head><title>416 Range not satisifiable</title></head>" +
		"<body>" +
		"<h1>Ooh dear</h1>" +
		"<p>Sorry, the file isn't that big. Maybe try asking for a bit before the end of the file?</p>" +
		"</body></html>");
	console.error("416 Bad Range - " + (e ? e.message : ""));
});

handler.on("security", function (res, e) {
	errorHeader(res, 403);
	res.end("<!DOCTYPE html><html lang=\"en\">" +
		"<head><title>403 Forbidden</title></head>" +
		"<body>" +
		"<h1>Hey!</h1>" +
		"<p>Stop trying to hack my server!</p>" +
		"</body></html>");
	console.error("403 Security - " + (e ? e.message : ""));
});

handler.on("badMime", function (res, e) {
	errorHeader(res, 403);
	res.end("<!DOCTYPE html><html lang=\"en\">" +
		"<head><title>403 Forbidden</title></head>" +
		"<body>" +
		"<h1>Sorry&hellip;</h1>" +
		"<p>You're not allowed to download files of that type.</p>" +
		"</body></html>");
	console.error("403 Bad MIME - " + (e ? e.message : ""));
});

handler.on("badRequest", function (res, e) {
	errorHeader(res, 400);
	res.end("<!DOCTYPE html><html lang=\"en\">" +
		"<head><title>400 Bad request</title></head>" +
		"<body>" +
		"<h1>Wut?</h1>" +
		"<p>I couldn't understand that I'm afraid; the syntax appears malformed.</p>" +
		"</body></html>");
	console.error("400 Bad Request - " + (e ? e.message : ""));
});

handler.on("noRandomFiles", function (res, e) {
	errorHeader(res, 404);
	res.end("<!DOCTYPE html><html lang=\"en\">" +
		"<head><title>404 Not found</title></head>" +
		"<body>" +
		"<h1>Sorry&hellip;</h1>" +
		"<p>There don't appear to be any files of that type at all there.</p>" +
		"</body></html>");
	console.error("404 No Random Files - " + (e ? e.message : ""));
});

/*process.on('uncaughtException', function(e) {
	util.debug(e);
});*/

module.exports = vidStreamer;
